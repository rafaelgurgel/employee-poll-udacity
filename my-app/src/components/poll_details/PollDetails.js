import React from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { useParams, Navigate, useNavigate } from 'react-router-dom';
import { handleAnswer } from '../../actions/shared';
import getAvatar from '../../utils/getAvatar';

import './poll_details.css';

export default function PollDetails() {
  const { id } = useParams();
  const navigate = useNavigate();
  const dispatch = useDispatch();

  const { question, author, user } = useSelector((state) => {
    const question = state.questions[id];
    const author = question ? state.users[question.author] : null;
    const authedUser = state.authedUser;
    const user = authedUser ? state.users[authedUser] : null;
    return { question, author, user };
  });

  
  if (!question) {
    return <Navigate to="/notfound" />;
  }
  if (!author || !user) {
    return <p>Required data not found</p>;
  }

  const hasVotedFor = user.answers[id];
  const totalVotes =
    question.optionOne.votes.length + question.optionTwo.votes.length;

  const handleVote = (answer) => {
    dispatch(handleAnswer(id, answer));
  };

  return (
    <div className="polldetails-container">
      <h3 className="polldetails-title">Poll by {author.name}</h3>
      <img
        src={getAvatar(author.avatarURL)}
        alt={`Avatar of ${author.name}`}
        className="polldetails-avatar"
      />

      <h4 className="polldetails-subtitle">Would You Rather...</h4>

      {hasVotedFor ? (
        <div>
          <div
            className="polldetails-results"
            style={
              hasVotedFor === 'optionOne'
                ? { fontWeight: 'bold', color: 'green' }
                : {}
            }
          >
            {question.optionOne.text}
            <br />
            {question.optionOne.votes.length} vote(s) (
            {((question.optionOne.votes.length / totalVotes) * 100).toFixed(2)}
            %)
          </div>
          <div
            className="polldetails-results"
            style={
              hasVotedFor === 'optionTwo'
                ? { fontWeight: 'bold', color: 'green' }
                : {}
            }
          >
            {question.optionTwo.text}
            <br />
            {question.optionTwo.votes.length} vote(s) (
            {((question.optionTwo.votes.length / totalVotes) * 100).toFixed(2)}
            %)
          </div>
        </div>
      ) : (
        <div className="polldetails-vote-buttons">
          <button onClick={() => handleVote('optionOne')}>
            {question.optionOne.text}
          </button>
          <button onClick={() => handleVote('optionTwo')}>
            {question.optionTwo.text}
          </button>
        </div>
      )}

      <br />
      <button
        className="polldetails-back"
        onClick={() => navigate('/home')}
      >
        Back
      </button>
    </div>
  );
}
