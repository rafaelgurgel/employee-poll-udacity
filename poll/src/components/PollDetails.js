import React from 'react'
import { useSelector, useDispatch } from 'react-redux'
import { useParams, useNavigate } from 'react-router-dom'
import { handleAnswer } from '../actions/shared'

function getAvatar(avatarURL) {
  return avatarURL ? avatarURL : 'https://via.placeholder.com/50'
}

export default function PollDetails() {
  const { id } = useParams();
  const navigate = useNavigate();
  const dispatch = useDispatch();

  const { question, author, user, authedUser } = useSelector((state) => {
    const question = state.questions[id];
    const author = question ? state.users[question.author] : null;
    const authedUser = state.authedUser;
    const user = authedUser ? state.users[authedUser] : null;
    return { question, author, user, authedUser };
  });

  if (!question) {
    return <p>404 - Poll not found</p>;
  }

  if (!author || !user) {
    return <p>Required data not found</p>;
  }

  const hasVotedFor = user.answers[id];
  const totalVotes = question.optionOne.votes.length + question.optionTwo.votes.length;

  const handleVote = (answer) => {
    dispatch(handleAnswer(id, answer));
  };

  return (
    <div style={{ marginTop: '1rem' }}>
      <h3>Poll by {author.name}</h3>
      <img
        src={getAvatar(author.avatarURL)}
        alt={`Avatar of ${author.name}`}
        width="50"
        height="50"
        style={{ borderRadius: '50%' }}
      />
      <h4>Would You Rather...</h4>
      {hasVotedFor ? (
        <div>
          <div
            style={
              hasVotedFor === 'optionOne'
                ? { fontWeight: 'bold', color: 'green' }
                : {}
            }
          >
            {question.optionOne.text} <br />
            {question.optionOne.votes.length} vote(s) (
            {((question.optionOne.votes.length / totalVotes) * 100).toFixed(2)}%)
          </div>
          <hr />
          <div
            style={
              hasVotedFor === 'optionTwo'
                ? { fontWeight: 'bold', color: 'green' }
                : {}
            }
          >
            {question.optionTwo.text} <br />
            {question.optionTwo.votes.length} vote(s) (
            {((question.optionTwo.votes.length / totalVotes) * 100).toFixed(2)}%)
          </div>
        </div>
      ) : (
        <div>
          <button onClick={() => handleVote('optionOne')}>
            {question.optionOne.text}
          </button>
          <button onClick={() => handleVote('optionTwo')}>
            {question.optionTwo.text}
          </button>
        </div>
      )}
      <br />
      <button onClick={() => navigate('/home')}>Back</button>
    </div>
  );
}
