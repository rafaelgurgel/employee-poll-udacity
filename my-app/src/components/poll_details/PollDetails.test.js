import React from 'react'
import { render } from '@testing-library/react'
import { Provider } from 'react-redux'
import { createStore } from 'redux'
import { BrowserRouter } from 'react-router-dom'
import PollDetails from './PollDetails'
import '@testing-library/jest-dom'

// For useParams mocking (React Router v6). We can do a jest mock or pass it via context:
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useParams: () => ({ id: 'q1' }),
  useNavigate: () => jest.fn(),
}))

const mockReducer = (state = {
  authedUser: 'sarahedo',
  questions: {
    q1: {
      id: 'q1',
      author: 'sarahedo',
      optionOne: { votes: [], text: 'OptionOne' },
      optionTwo: { votes: ['sarahedo'], text: 'OptionTwo' }
    },
  },
  users: {
    sarahedo: { id: 'sarahedo', name: 'Sarah Edo', answers: { q1: 'optionTwo' }, avatarURL: null },
  }
}) => state

const mockStore = createStore(mockReducer)

describe('PollDetails Component', () => {
  it('matches snapshot', () => {
    const { asFragment } = render(
      <Provider store={mockStore}>
        <BrowserRouter>
          <PollDetails />
        </BrowserRouter>
      </Provider>
    )
    expect(asFragment()).toMatchSnapshot()
  })
})
