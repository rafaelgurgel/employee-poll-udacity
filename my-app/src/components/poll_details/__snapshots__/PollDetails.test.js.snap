// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PollDetails Component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="polldetails-container"
  >
    <h3
      class="polldetails-title"
    >
      Poll by Sarah Edo
    </h3>
    <img
      alt="Avatar of Sarah Edo"
      class="polldetails-avatar"
      src="defaultAvatar.svg"
    />
    <h4
      class="polldetails-subtitle"
    >
      Would You Rather...
    </h4>
    <div>
      <div
        class="polldetails-results"
      >
        OptionOne
        <br />
        0 vote(s) (0.00%)
      </div>
      <div
        class="polldetails-results"
        style="font-weight: bold; color: green;"
      >
        OptionTwo
        <br />
        1 vote(s) (100.00%)
      </div>
    </div>
    <br />
    <button
      class="polldetails-back"
    >
      Back
    </button>
  </div>
</DocumentFragment>
`;
